LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs

```diff
@@ -24,79 +21,146 @@ namespace LoRaWan.NetworkServer.BasicsStation.Processors
+#pragma warning restore CA1508 // Avoid dead conditional code
+            {
+                // This can happen if the basic station client is losing connectivity
+                this.logger.LogDebug(ex, ex.Message);
+            }
+
+            return httpContext;
         }
 
+        public Task HandleDiscoveryAsync(HttpContext httpContext, CancellationToken cancellationToken) =>
+            ProcessIncomingRequestAsync(httpContext, InternalHandleDiscoveryAsync, cancellationToken);
+
+        public Task HandleDataAsync(HttpContext httpContext, CancellationToken cancellationToken) =>
+            ProcessIncomingRequestAsync(httpContext,
+                                        (httpContext, socket, ct) => InternalHandleDataAsync(httpContext.Request.RouteValues, socket, ct),
+                                        cancellationToken);
+
         /// <returns>A boolean stating if more requests are expected on this endpoint. If false, the underlying socket should be closed.</returns>
-        internal async Task<bool> InternalHandleDiscoveryAsync(HttpContext httpContext, string json, WebSocket socket, CancellationToken token)
+        internal async Task InternalHandleDiscoveryAsync(HttpContext httpContext, WebSocket socket, CancellationToken cancellationToken)
         {
-            var stationEui = LnsDiscovery.QueryReader.Read(json);
-            this.logger.LogInformation($"Received discovery request from: {stationEui}");
-
-            try
+            await using var message = socket.ReadTextMessages(cancellationToken);
+            if (!await message.MoveNextAsync())
+            {
+                this.logger.LogWarning($"Did not receive discovery request from station.");
+            }
+            else
             {
-                var scheme = httpContext.Request.IsHttps ? "wss" : "ws";
-                var url = new Uri($"{scheme}://{httpContext.Request.Host}{BasicsStationNetworkServer.DataEndpoint}/{stationEui}");
+                var json = message.Current;
+                var stationEui = LnsDiscovery.QueryReader.Read(json);
+                this.logger.LogInformation($"Received discovery request from: {stationEui}");
 
-                var networkInterface = NetworkInterface.GetAllNetworkInterfaces()
-                                                       .SingleOrDefault(ni => ni.GetIPProperties()
+                try
+                {
+                    var scheme = httpContext.Request.IsHttps ? "wss" : "ws";
+                    var url = new Uri($"{scheme}://{httpContext.Request.Host}{BasicsStationNetworkServer.DataEndpoint}/{stationEui}");
+
+                    var networkInterface = NetworkInterface.GetAllNetworkInterfaces()
+                                                           .SingleOrDefault(ni => ni.GetIPProperties()
                                                                                 .UnicastAddresses
```

Comment:
nit: these rows seem misaligned to me 👍

Reply:
Indeed. Fixed with 889fb2ce5d4ac3c429c4f59013f670d20aee7a7a.